### PR TITLE
Added UberAuthClient and AuthActivity

### PR DESCRIPTION
**Pull request by lalwani**
Provided entry point for the Uber auth sdk to initiate the authentication flow.
Added AuthActivity for a couple of reasons

- have control over onActivityResult callback to perform PKCE if required
- Use the new api (registerActivityForResult) for fetching activity result from other uber apps within the sdk
- to provide any UX (not today) when initiating the flow

### CODE REVIEW

File: authentication/lint.xml [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+  ~
+  ~ Unless required by applicable law or agreed to in writing, software
+  ~ distributed under the License is distributed on an "AS IS" BASIS,
+  ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  ~ See the License for the specific language governing permissions and
+  ~ limitations under the License.
+-->
+
+<lint>
+    <issue id="InvalidPackage">
```
Comment (tyvsmith) [inline authentication/lint.xml:20]: Why is this needed?

File: authentication/src/main/kotlin/com/uber/sdk2/auth/client/UberAuthClient.kt [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import android.content.Intent
+import androidx.activity.result.ActivityResultLauncher
+import com.uber.sdk2.auth.AuthActivity
+import com.uber.sdk2.auth.api.request.AuthContext
+
+/**
+ * Client to authenticate the user against one of the available Uber apps. It is entrypoint for the
+ * consumers to initiate the authentication flow
+ */
+object UberAuthClient {
```
Comment (psteiger) [inline authentication/src/main/kotlin/com/uber/sdk2/auth/client/UberAuthClient.kt:29]: Do we want this to be an `interface` so we can provide alternative impl? (example: tests)

File: authentication/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -17,4 +17,39 @@
 
 <manifest xmlns:android="http://schemas.android.com/apk/res/android">
 
+    <uses-permission android:name="android.permission.INTERNET" />
+
+    <queries>
+        <package android:name="com.ubercab" />
+        <package android:name="com.ubercab.presidio.exo" />
```
Comment (tyvsmith) [inline authentication/src/main/AndroidManifest.xml:24]: Do we want all 3rd party apps to work with dev builds?

**Review by psteiger — APPROVED** [review]

File: authentication/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -17,4 +17,39 @@
 
 <manifest xmlns:android="http://schemas.android.com/apk/res/android">
 
+    <uses-permission android:name="android.permission.INTERNET" />
+
+    <queries>
+        <package android:name="com.ubercab" />
```
Comment (tyvsmith) [inline authentication/src/main/AndroidManifest.xml:23]: Are we doing fingerprint validation against these before talking to them? We should be. If that's already in place somewhere else, please point me to it, if not, can you have me be an explicit reviewer on that?

Comment (lalwani) [reply authentication/src/main/AndroidManifest.xml:23]: We are invoking the Uber apps via applinks. If the Uber app is not registered for that applink (via assetlinks.json) it will not be invoked so fingerprint validation is not required.

File: authentication/src/main/AndroidManifest.xml (hunk 3 above)
Comment (lalwani) [reply authentication/src/main/AndroidManifest.xml:24]: We supported these with the old sdk so I carried them over (maybe for dev purpose?). I'm ok with just restricting them to prod and beta builds.

File: authentication/src/main/kotlin/com/uber/sdk2/auth/AuthActivity.kt [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        if (authCode.isNullOrEmpty()) {
+          throw AuthException.ClientError(AUTH_CODE_INVALID)
+        }
+        authProvider?.handleAuthCode(authCode)
+      }
+    }
+  }
+
+  private fun isAuthorizationCodePresent(uri: Uri): Boolean {
+    return !TextUtils.isEmpty(uri.getQueryParameter(KEY_AUTHENTICATION_CODE))
```
Comment (psteiger) [inline authentication/src/main/kotlin/com/uber/sdk2/auth/AuthActivity.kt:94]: just use Kotlin's `isEmpty()` instead

File: authentication/lint.xml (hunk 1 above)
Comment (lalwani) [reply authentication/lint.xml:20]: Followed same lint rules as the old sdk

File: authentication/src/main/kotlin/com/uber/sdk2/auth/AuthActivity.kt [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+
+    authProvider = AuthProvider(this, authContext)
+    init()
+  }
+
+  private fun init() {
+    authProvider?.let {
+      lifecycleScope.launch(Dispatchers.IO) {
+        val authResult = it.authenticate()
+        withContext(Dispatchers.Main) {
```
Comment (psteiger) [inline authentication/src/main/kotlin/com/uber/sdk2/auth/AuthActivity.kt:54]: why dont we use `withContext()` in `authenticate()` instead?

Method should be safe to call from any context.

File: authentication/src/main/kotlin/com/uber/sdk2/auth/client/UberAuthClient.kt (hunk 2 above)
Comment (lalwani) [reply authentication/src/main/kotlin/com/uber/sdk2/auth/client/UberAuthClient.kt:29]: The rationale behind object was to provide a single yet easy way to launch the auth flow. 
But, your raise a good point about testability. I'll update with a new patch

File: authentication/src/main/AndroidManifest.xml (hunk 4 above)
Comment (lalwani) [reply authentication/src/main/AndroidManifest.xml:23]: * verified for that applink

File: authentication/src/main/kotlin/com/uber/sdk2/auth/internal/AuthActivity.kt [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+      intent.getParcelableExtra<AuthContext>(AUTH_CONTEXT)
+        ?: throw IllegalStateException("AuthContext is required")
+
+    authProvider = AuthProvider(this, authContext)
+    init()
+  }
+
+  private fun init() {
+    authProvider?.let {
+      lifecycleScope.launch(Dispatchers.IO) {
```
Comment (psteiger) [inline authentication/src/main/kotlin/com/uber/sdk2/auth/internal/AuthActivity.kt:49]: I think this is not right. We can't call `setResult` and `finish()` on `IO` context.

I think we should use `lifecycleScope.launch { }` (so, `Main`) and just switch context inside `authenticate()` *if needed*

Comment (lalwani) [reply authentication/src/main/kotlin/com/uber/sdk2/auth/internal/AuthActivity.kt]: yea, figured I could just invert it and don't need IO dispatcher for retrofit call either since the methods are already marked as `suspend`
